staking-pool-shares/src/lib.rs

```diff
@@ -0,0 +1,460 @@
+        // amount.
+        // Rounding up `a / b` using `(a + b - 1) / b`.
+        let num_shares = ((U256::from(self.total_staked_shares) * U256::from(amount)
+            + U256::from(self.total_staked_balance - 1))
+            / U256::from(self.total_staked_balance))
+        .as_u128();
+
+        assert!(
+            account.staked_shares >= num_shares,
+            "Not enough staked balance to unstake"
+        );
+        account.staked_shares -= num_shares;
+        account.unstaked += amount;
+        account.unstaked_available_epoch_height = env::epoch_height() + EPOCHS_TOWARDS_REWARD;
+        self.save_account(&account_id, account);
+
+        self.total_staked_balance -= amount;
+        self.total_staked_shares -= num_shares;
+
+        self.restake();
+    }
+
+    fn restake(&mut self) {
+        Promise::new(env::current_account_id())
+            .function_call(b"ping".to_vec(), b"{}".to_vec(), 0, PING_GAS)
+            .stake(self.total_staked_balance, self.stake_public_key.clone())
+            .function_call(
+                b"internal_after_stake".to_vec(),
+                b"{}".to_vec(),
+                0,
+                REFRESH_STAKE_GAS,
+            );
+    }
+
+    /// Private method to be called after stake action.
+    pub fn internal_after_stake(&mut self) {
+        assert_eq!(env::current_account_id(), env::predecessor_account_id());
+        self.last_account_balance = env::account_balance();
+        self.last_locked_account_balance = env::account_locked_balance();
+    }
+
+    /// Returns given account's unstaked balance.
+    pub fn get_account_unstaked_balance(&self, account_id: AccountId) -> U128 {
+        self.get_account(&account_id).unstaked.into()
+    }
+
+    /// Returns given account's staked balance.
+    pub fn get_account_staked_balance(&self, account_id: AccountId) -> U128 {
+        if self.total_staked_shares > 0 {
+            // Rounding down
+            (U256::from(self.total_staked_balance)
+                * U256::from(self.get_account(&account_id).staked_shares)
+                / U256::from(self.total_staked_shares))
+            .as_u128()
+            .into()
+        } else {
+            0.into()
+        }
+    }
+
+    pub fn get_account_total_balance(&self, account_id: AccountId) -> U128 {
+        (self.get_account(&account_id).unstaked + self.get_account_staked_balance(account_id).0)
+            .into()
+    }
+
+    pub fn is_account_unstaked_balance_available(&self, account_id: AccountId) -> bool {
+        self.get_account(&account_id)
+            .unstaked_available_epoch_height
+            <= env::epoch_height()
+    }
+
+    fn get_account(&self, account_id: &AccountId) -> Account {
+        self.accounts.get(account_id).unwrap_or_default()
+    }
+
+    fn save_account(&mut self, account_id: &AccountId, account: Account) {
+        if account != Account::default() {
```

Comment:
take account: Option<Account> and on None delete?

Reply:
This is easier to manage. It's basically clears the map when the account is default. But actually I need to only compare `unstaked` and `staked_shares` to 0, and ignore `epoch_height`